src/plugins/explore/public/components/query_panel/components/editor_stack/resuable_editor.tsx

```diff
@@ -0,0 +1,184 @@
+        blurTimeoutRef.current = setTimeout(() => {
+          setEditorIsFocused(false);
+        }, 300);
+      });
+
+      editorRef.current = editor;
+
+      editor.addAction({
+        id: `run-on-enter-${editorType}`,
+        label: `Run on ${editorType} Enter`,
+        keybindings: [monaco.KeyCode.Enter],
+        contextMenuGroupId: 'navigation',
+        contextMenuOrder: 1.5,
+        run: () => {
+          const val = editor.getValue();
+          onRun(val);
+          onEdit();
+        },
+      });
+
+      editor.addAction({
+        id: `insert-new-line-${editorType}`,
+        label: `Insert New Line on ${editorType}`,
+        // eslint-disable-next-line no-bitwise
+        keybindings: [monaco.KeyMod.Shift | monaco.KeyCode.Enter],
+        run: (ed) => {
+          if (ed.hasTextFocus()) {
+            const currentPosition = ed.getPosition();
+            if (currentPosition) {
+              ed.executeEdits('', [
+                {
+                  range: new monaco.Range(
+                    currentPosition.lineNumber,
+                    currentPosition.column,
+                    currentPosition.lineNumber,
+                    currentPosition.column
+                  ),
+                  text: '\n',
+                  forceMoveMarkers: true,
+                },
+              ]);
+              ed.setPosition({
+                lineNumber: currentPosition.lineNumber + 1,
+                column: 1,
+              });
+            }
+          }
+        },
+      });
+
+      editor.onDidContentSizeChange(() => {
+        const contentHeight = editor.getContentHeight();
+        setEditorHeight(contentHeight);
+        editor.layout({ width: editor.getLayoutInfo().width, height: contentHeight });
+      });
+
+      onEditorDidMount?.(editor);
+
+      return () => {
+        focusDisposable.dispose();
+        blurDisposable.dispose();
+        if (blurTimeoutRef.current) clearTimeout(blurTimeoutRef.current);
+        return editor;
+      };
+    },
+    [onRun, onEdit, editorType, onEditorDidMount]
+  );
+
+  return (
+    <div
+      className={`${editorType}EditorWrapper`}
+      style={{
+        position: 'relative',
+      }}
+    >
+      <div
+        className={`${editorType}Editor ${isReadOnly ? `${editorType}Editor--readonly` : ''}`}
+        style={{
+          borderBottom: editorIsFocused && !isReadOnly ? '1px solid #006BB4' : undefined,
+        }}
+        data-test-subj={`osd${editorType}Editor__multiLine`}
+      >
+        <CodeEditor
+          height={editorHeight}
+          languageId={editorConfig.languageId}
+          value={value}
+          onChange={onChange}
+          editorDidMount={handleEditorDidMount}
+          options={editorConfig.options}
+          languageConfiguration={editorConfig.languageConfiguration}
+          triggerSuggestOnFocus={editorConfig.triggerSuggestOnFocus}
+        />
+
+        {!value && !editorIsFocused && !isReadOnly && (
```

Comment:
can we encapsulate this to a variable. like `shouldDisplayPlaceholder` or something